torch/utils/data/distributed.py

```diff
@@ -61,7 +73,14 @@ def __iter__(self):
         return iter(indices)
 
     def __len__(self):
-        return self.num_samples
+        if self.padding:
+            return self.num_samples
+        return self._dataset_length
+
+    def set_rank(self, rank):
```

Comment:
This seems like a bigger change to how the sampler is used. Outside of your intended use-case, how is this useful and why is this necessary?

Reply:
After consolidating all feedbacks to eliminate `DistributedChunkSampler` from https://github.com/pytorch/pytorch/pull/26547 and after doing local tests, I noticed that we need this setter along with the other changes in this PR.

In a nutshell, for distributed training using `ChunkDataset` datasets with multiple dataloader workers, every process will have a copy of the dataset, which has a copy of the `DistributedSampler`. The initial value of sampler's `rank` must be `None` as it is the `DataLoader` who knows how many `num_workers` there will be. So we need a setter so that each process can assign their rank ID to the sampler (https://github.com/pytorch/pytorch/blob/8362a13a4e9791923ac5967506dc41a9a24e5be7/torch/utils/data/dataset.py#L373 and https://github.com/pytorch/pytorch/blob/8362a13a4e9791923ac5967506dc41a9a24e5be7/torch/utils/data/dataset.py#L327 shows their use)

Reply:
You're passing in the sampler during construction of the DataLoader I assume (since DataLoader takes a sampler argument), so you will know the number of workers. Why not use that information?

Reply:
We cannot pass any sampler to the `DataLoader` when `IterableDataset` is used - by design. That is actually the reason why we need an external sampler and the ability to set rank after DataLoader spawns workers

Yes, `set_rank` allows workers to register themselves after DataLoader creates them. In a similar fashion, the existing `set_epoch` method allows the training loop to register the current epoch with the sampler too.

Reply:
Sounds like we should be able to add a sampler to DataLoader. For example, we might want to suppor reservoir samplers. Let's change the DataLoader to support this or figure out why it can't.

Reply:
See https://github.com/pytorch/pytorch/issues/28743

Reply:
`DataLoader`'s sampler was originally meant to return example indices, which works great for map-style datasets with known size.

After `IterableDataset` was introduced, on the other hand, `DataLoader` doesnt  know the size of the iterable dataset and it must iterate over it until `StopIteration` is raised. In this case, it makes sense to disable external samplers and create an internal sampler, but at the same time a gap with sampling was created.

In order to add back `sampler` to `DataLoader`, we would have to make it clear that the sampler will be used by `IterableDataset` in the worker process as opposed to be used by `DataLoader` (as it would be if dataset was map-style).

In both cases (modifying `DistributedSampler` or changing `DataLoader` to accept sampler for IterableDataset), in order to prevent `DataLoader` to be re-instantiated every epoch, destroying all workers context, a `set_rank` would be needed so that worker processes can use it.

Therefore, IMHO, this PR is still valid as `padding`, `length` and `set_rank` would be used by both approaches: `Dataloader.sampler` being used by `IterableDataset` or `IterableDataset.sample` being explicitly passed.

Reply:
Ok makes sense. Then set_rank should also work for https://github.com/pytorch/pytorch/issues/28743 (I saw you already commented). Maybe that's a bit beyond scope, but if we implement another sampler we could see if set_rank is a general, orthogonal extension we need for iterator-style samplers. For example some simple  sampling algorithm such as a variation of https://en.wikipedia.org/wiki/Reservoir_sampling

Reply:
`set_rank` can be used for reservoir sampling when reading duplicate examples in different workers are not desired. Some changes on the original algorithm might be necessary to make reservoir sampling fully distributed among workers, but that is a bit out of scope indeed :)

Using reservoir sampling could the next use case for https://github.com/pytorch/pytorch/pull/26547 once it is merged, as all the boiler plate code to distribute training with embedded sampler on iterable datasets will already be there. The original PR handles known chunk size, but if implement a reservoir sampler, `ChunkDataset` can leverage it to train on unlimited number of chunks